core/src/main/java/io/grpc/ServiceProviders.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2017, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ServiceConfigurationError;
+import java.util.ServiceLoader;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+final class ServiceProviders {
+  private static final Logger logger = Logger.getLogger(ServiceProviders.class.getName());
+
+  private ServiceProviders() {
+    // do not instantiate
+  }
+
+  /**
+   * If this is not Android, returns the highest priority implementation of the class via
+   * {@link ServiceLoader}.
+   * If this is Android, returns the highest priority class in the hardcoded
+   * {@code androidClassses}.
+   */
+  public static <T extends ServiceProvider> T load(
+      Class<T> klass, List<String> androidClasses, ClassLoader cl) {
+    List<T> candidates = loadAll(klass, androidClasses, cl);
+    if (candidates.isEmpty()) {
+      return null;
+    }
+    return candidates.get(0);
+  }
+
+  /**
+   * If this is not Android, returns all available implementations discovered via
+   * {@link ServiceLoader}.
+   * If this is Android, returns all available implementations in the hardcoded
+   * {@code androidClasses}.
+   * The list is sorted in descending priority order.
+   */
+  public static <T extends ServiceProvider> List<T> loadAll(
+      Class<T> klass, List<String> androidClasses, ClassLoader cl) {
+    Iterable<T> candidates;
+    if (isAndroid(cl)) {
+      candidates = getCandidatesViaHardCoded(klass, androidClasses, cl);
+    } else {
+      candidates = getCandidatesViaServiceLoader(klass, cl);
+    }
+    List<T> list = new ArrayList<T>();
+    Iterator<T> iter = candidates.iterator();
+    while (iter.hasNext()) {
+      try {
+        T current = iter.next();
+        if (!current.isAvailable()) {
+          continue;
+        }
+        list.add(current);
+      } catch (Throwable t) {
+        // The iterator from ServiceLoader may throw ServiceConfigurationError
```

Comment:
This was purposefully being propagated before. I don't think we can safely continue. Different implementations have different behavior, so continuing when the user expected a certain implementation seems dangerous.

Note that this catch is also incomplete, because `hasNext()` can throw.

Reply:
This makes sense, I'll change it so that we just give up if any classes fail.